task/mgmt-gateway/src/update/host_flash.rs

```diff
@@ -0,0 +1,315 @@
+
+        // What is the total capacity of the device?
+        let capacity = self
+            .task
+            .capacity()
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+
+        // How many total sectors do we need to erase? For gimlet, we know that
+        // capacity is an exact multiple of the sector size, which is probably
+        // a safe assumption for future parts as well. We'll assert here in case
+        // that ever becomes untrue, and we can update our math.
+        assert!(capacity % SECTOR_SIZE_BYTES == 0);
+        let num_sectors = (capacity / SECTOR_SIZE_BYTES) as u32;
+
+        self.current = Some(CurrentUpdate::new(
+            update.id,
+            update.total_size,
+            State::ErasingSectors {
+                buffer,
+                sectors_to_erase: 0..num_sectors,
+            },
+        ));
+
+        Ok(())
+    }
+
+    fn is_preparing(&self) -> bool {
+        match self.current.as_ref().map(CurrentUpdate::state) {
+            Some(State::ErasingSectors { .. }) => true,
+            Some(State::AcceptingData { .. })
+            | Some(State::Complete)
+            | Some(State::Failed(_))
+            | Some(State::Aborted)
+            | None => false,
+        }
+    }
+
+    fn step_preparation(&mut self) {
+        let current = match self.current.as_mut() {
+            Some(current) => current,
+            None => return,
+        };
+
+        current.update_state(|state| {
+            let (buffer, mut sectors_to_erase) = match state {
+                State::ErasingSectors {
+                    buffer,
+                    sectors_to_erase,
+                } => (buffer, sectors_to_erase),
+                State::AcceptingData { .. }
+                | State::Complete
+                | State::Failed(_)
+                | State::Aborted => {
+                    // Nothing to prepare in any of these states; put it back.
+                    return state;
+                }
+            };
+
+            let addr = sectors_to_erase.start * SECTOR_SIZE_BYTES as u32;
+            match self.task.sector_erase(addr as u32) {
+                Ok(()) => {
+                    sectors_to_erase.start += 1;
+                    if sectors_to_erase.start == sectors_to_erase.end {
+                        State::AcceptingData {
+                            buffer,
+                            next_write_offset: 0,
+                        }
+                    } else {
+                        State::ErasingSectors {
+                            buffer,
+                            sectors_to_erase,
+                        }
+                    }
+                }
+                Err(err) => State::Failed(err),
+            }
+        });
+    }
+
+    fn status(&self) -> UpdateStatus {
+        let current = match self.current.as_ref() {
+            Some(current) => current,
+            None => return UpdateStatus::None,
```

Comment:
Would it make sense to remove `UpdateStatus::None` and instead return an `Option<UpdateStatus>`?  

(Not at all necessary for this PR, since it would require `omicron` changes as well)

Reply:
That's what it was originally, but in PR review on the `omicron` side we switched it for clarity / readability reasons (e.g., when skimming code, we always have an `UpdateStatus` available, but one of the possible status values is "no update in progress"). A minuscule technical benefit is that the hubpack encoding of `UpdateStatus` (with a `None` case) is one byte shorter than the encoding of `Option<UpdateStatus>`.